File: src/messenger/discord_adapter.py
"""
Discord adapter for Solo Founder Agents.
"""

from __future__ import annotations

import os
import json
from pathlib import Path
from typing import Callable, Awaitable

import discord
from discord.ext import commands
from dotenv import load_dotenv

from .base import MessengerAdapter, MessageContext

load_dotenv()

TOKEN = os.getenv("DISCORD_TOKEN")
REPOS_BASE = Path(os.getenv("REPOS_BASE_PATH", "./repos"))
PRODUCTS_FILE = Path("core/products.json")


def _load_products() -> list[dict]:
    if PRODUCTS_FILE.exists():
        return json.loads(PRODUCTS_FILE.read_text())
    return []


class DiscordMessageContext(MessageContext):
    def __init__(self, message: discord.Message, product: dict, agent_label: str = ""):
        self._message = message
        self._product = product
        self._agent_label = agent_label

    async def reply(self, text: str) -> None:
        chunks = [text[i : i + 1900] for i in range(0, len(text), 1900)]
        for i, chunk in enumerate(chunks):
            prefix = ""
            if i == 0:
                prefix = f"**[{self._product['name']}{self._agent_label}]**\n"
            await self._message.reply(f"{prefix}```\n{chunk}\n```")

    async def typing(self) -> None:
        await self._message.channel.trigger_typing()


class DiscordAdapter(MessengerAdapter):
    """Discord implementation using discord.py."""

    def __init__(self):
        self._bot: commands.Bot | None = None
        self._client: discord.Client | None = None

    @property
    def platform(self) -> str:
        return "discord"

    # ── Bot mode ──

    async def start_bot(
        self,
        on_command: Callable[[str, dict, MessageContext], Awaitable[None]],
    ) -> None:
        intents = discord.Intents.default()
        intents.message_content = True
        intents.guilds = True

        bot = commands.Bot(command_prefix="!", intents=intents)
        self._bot = bot

        @bot.event
        async def on_ready():
            print(f"[Discord Bot] Logged in: {bot.user}")
            for guild in bot.guilds:
                await self._ensure_channels(guild)
            await self._sync_guild_product_mapping()
            print(f"[Discord Bot] Ready. Connected to {len(bot.guilds)} server(s)")

        @bot.event
        async def on_guild_join(guild: discord.Guild):
            await self._ensure_channels(guild)
            await self._sync_guild_product_mapping()

        @bot.event
        async def on_message(message: discord.Message):
            if message.author.bot:
                return
            if message.channel.name != "owner-command":
                return

            product = self._get_product_by_guild(message.guild.id)
            if not product:
                await message.channel.send(
                    "No product linked to this server. Re-run `setup.py`."
                )
                return

            ctx = DiscordMessageContext(message, product)
            async with message.channel.typing():
                await on_command(message.content.strip(), product, ctx)
            await bot.process_commands(message)

        if not TOKEN:
            print("DISCORD_TOKEN is not set. Check your .env file.")
            raise SystemExit(1)

        await bot.start(TOKEN)

    # ── Notifier mode ──

    async def start_notifier(self) -> None:
        intents = discord.Intents.default()
        self._client = discord.Client(intents=intents)

        if not TOKEN:
            print("DISCORD_TOKEN is not set.")
            raise SystemExit(1)

    async def send_to_channel(
        self,
        product_config: dict,
        channel_name: str,
        text: str,
        title: str | None = None,
    ) -> bool:
        client = self._client or self._bot
        if not client or not client.is_ready():
            return False

        guild_id = product_config.get("guild_id")
        if not guild_id:
            return False

        channel_key = channel_name.replace("-", "_")
        channel_id = product_config.get("channels", {}).get(channel_key)
        if not channel_id:
            return False

        channel = client.get_channel(channel_id)
        if not channel:
            return False

        content = f"**{title}**\n\n{text}" if title else text
        for chunk in [content[i : i + 1900] for i in range(0, len(content), 1900)]:
            await channel.send(chunk)
        return True

    async def setup_channels(self, product_config: dict) -> list[str]:
        client = self._client or self._bot
        if not client:
            return []

        guild_id = product_config.get("guild_id")
        if not guild_id:
            return []

        guild = client.get_guild(guild_id)
        if not guild:
            return []

        return await self._ensure_channels(guild)

    # ── Internal helpers ──

    async def _ensure_channels(self, guild: discord.Guild) -> list[str]:
        existing = {ch.name for ch in guild.channels}

        category = discord.utils.get(guild.categories, name="AI Team Reports")
        if not category:
            category = await guild.create_category("AI Team Reports")

        created = []
        for ch_name in self.channel_names:
            if ch_name not in existing:
                await guild.create_text_channel(ch_name, category=category)
                created.append(ch_name)

        if created:
            print(f"[Discord] {guild.name}: channels created → {', '.join(created)}")
        return created

    async def _sync_guild_product_mapping(self):
        import yaml

        bot = self._bot
        if not bot:
            return

        products = _load_products()
        for guild in bot.guilds:
            for product in products:
                if (
                    product["name"] in guild.name
                    or product["slug"] in guild.name.lower()
                ):
                    config_file = (
                        REPOS_BASE / product["slug"] / "discord" / "config.yaml"
                    )
                    if config_file.exists():
                        config = yaml.safe_load(config_file.read_text())
                        if config.get("guild_id") != guild.id:
                            config["guild_id"] = guild.id
                            for ch in guild.channels:
                                if ch.name in self.channel_names:
                                    key = ch.name.replace("-", "_")
                                    config["channels"][key] = ch.id
                            config_file.write_text(
                                yaml.dump(config, allow_unicode=True)
                            )
                            print(
                                f"[Discord] Mapped: {guild.name} ↔ {product['name']}"
                            )

    def _get_product_by_guild(self, guild_id: int) -> dict | None:
        import yaml

        products = _load_products()
        for p in products:
            config_file = REPOS_BASE / p["slug"] / "discord" / "config.yaml"
            if config_file.exists():
                config = yaml.safe_load(config_file.read_text())
                if config.get("guild_id") == guild_id:
                    return p
        return None

    def get_client(self) -> discord.Client | None:
        """Return the underlying client (for scheduler to call .run())."""
        return self._client

File: src/messenger/discord-adapter.ts
import {
  Client,
  GatewayIntentBits,
  ChannelType,
  type Message,
  type Guild,
  type TextChannel,
} from "discord.js";
import fs from "fs";
import path from "path";
import yaml from "js-yaml";
import { getReposBase } from "../util/paths.js";
import { loadProducts, type Product } from "../util/config.js";
import {
  DEFAULT_CHANNELS,
  type MessengerAdapter,
  type MessageContext,
  type CommandHandler,
} from "./base.js";

class DiscordMessageContext implements MessageContext {
  _agentLabel = "";
  constructor(
    private message: Message,
    private product: Product
  ) {}

  async reply(text: string): Promise<void> {
    const chunks = text.match(/.{1,1900}/gs) || [text];
    for (let i = 0; i < chunks.length; i++) {
      const prefix = i === 0 ? `**[${this.product.name}${this._agentLabel}]**\n` : "";
      await this.message.reply(`${prefix}\`\`\`\n${chunks[i]}\n\`\`\``);
    }
  }

  async typing(): Promise<void> {
    if ("sendTyping" in this.message.channel) {
      await (this.message.channel as TextChannel).sendTyping();
    }
  }
}

export class DiscordAdapter implements MessengerAdapter {
  readonly platform = "discord";
  readonly channelNames = DEFAULT_CHANNELS;
  private client: Client | null = null;

  async startBot(onCommand: CommandHandler): Promise<void> {
    const token = process.env.DISCORD_TOKEN;
    if (!token) {
      console.log("DISCORD_TOKEN is not set. Check your .env file.");
      process.exit(1);
    }

    const client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
      ],
    });
    this.client = client;

    client.on("ready", async () => {
      console.log(`[Discord Bot] Logged in: ${client.user?.tag}`);
      for (const guild of client.guilds.cache.values()) {
        await this.ensureChannels(guild);
      }
      this.syncGuildProductMapping();
      console.log(`[Discord Bot] Ready. Connected to ${client.guilds.cache.size} server(s)`);
    });

    client.on("guildCreate", async (guild) => {
      await this.ensureChannels(guild);
      this.syncGuildProductMapping();
    });

    client.on("messageCreate", async (message) => {
      if (message.author.bot) return;
      if ((message.channel as TextChannel).name !== "owner-command") return;

      const product = this.getProductByGuild(message.guild!.id);
      if (!product) {
        await message.channel.send("No product linked to this server. Re-run `solo-agents init`.");
        return;
      }

      const ctx = new DiscordMessageContext(message, product);
      await message.channel.sendTyping();
      await onCommand(message.content.trim(), product, ctx);
    });

    await client.login(token);
  }

  async startNotifier(): Promise<void> {
    const token = process.env.DISCORD_TOKEN;
    if (!token) {
      console.log("DISCORD_TOKEN is not set.");
      process.exit(1);
    }
    this.client = new Client({ intents: [GatewayIntentBits.Guilds] });
    await this.client.login(token);
  }

  async sendToChannel(
    productConfig: Record<string, unknown>,
    channelName: string,
    text: string,
    title?: string
  ): Promise<boolean> {
    if (!this.client?.isReady()) return false;

    const guildId = productConfig.guild_id as string | undefined;
    if (!guildId) return false;

    const channelKey = channelName.replace(/-/g, "_");
    const channels = (productConfig.channels || {}) as Record<string, string>;
    const channelId = channels[channelKey];
    if (!channelId) return false;

    const channel = this.client.channels.cache.get(channelId) as TextChannel | undefined;
    if (!channel) return false;

    const content = title ? `**${title}**\n\n${text}` : text;
    const chunks = content.match(/.{1,1900}/gs) || [content];
    for (const chunk of chunks) {
      await channel.send(chunk);
    }
    return true;
  }

  async setupChannels(productConfig: Record<string, unknown>): Promise<string[]> {
    if (!this.client) return [];
    const guildId = productConfig.guild_id as string | undefined;
    if (!guildId) return [];
    const guild = this.client.guilds.cache.get(guildId);
    if (!guild) return [];
    return this.ensureChannels(guild);
  }

  getClient(): Client | null {
    return this.client;
  }

  // -- Internal --

  private async ensureChannels(guild: Guild): Promise<string[]> {
    const existing = new Set(guild.channels.cache.map((ch) => ch.name));

    let category = guild.channels.cache.find(
      (ch) => ch.type === ChannelType.GuildCategory && ch.name === "AI Team Reports"
    );
    if (!category) {
      category = await guild.channels.create({
        name: "AI Team Reports",
        type: ChannelType.GuildCategory,
      });
    }

    const created: string[] = [];
    for (const chName of this.channelNames) {
      if (!existing.has(chName)) {
        await guild.channels.create({
          name: chName,
          type: ChannelType.GuildText,
          parent: category.id,
        });
        created.push(chName);
      }
    }

    if (created.length) {
      console.log(`[Discord] ${guild.name}: channels created → ${created.join(", ")}`);
    }
    return created;
  }

  private syncGuildProductMapping(): void {
    if (!this.client) return;
    const products = loadProducts();
    const reposBase = getReposBase();

    for (const guild of this.client.guilds.cache.values()) {
      for (const product of products) {
        if (guild.name.includes(product.name) || guild.name.toLowerCase().includes(product.slug)) {
          const configFile = path.join(reposBase, product.slug, "discord", "config.yaml");
          if (!fs.existsSync(configFile)) continue;

          const config = yaml.load(fs.readFileSync(configFile, "utf-8")) as Record<string, unknown>;
          if (config.guild_id !== guild.id) {
            config.guild_id = guild.id;
            const channels = (config.channels || {}) as Record<string, string>;
            for (const ch of guild.channels.cache.values()) {
              if (this.channelNames.includes(ch.name)) {
                channels[ch.name.replace(/-/g, "_")] = ch.id;
              }
            }
            config.channels = channels;
            fs.writeFileSync(configFile, yaml.dump(config));
            console.log(`[Discord] Mapped: ${guild.name} ↔ ${product.name}`);
          }
        }
      }
    }
  }

  private getProductByGuild(guildId: string): Product | null {
    const products = loadProducts();
    const reposBase = getReposBase();

    for (const p of products) {
      const configFile = path.join(reposBase, p.slug, "discord", "config.yaml");
      if (!fs.existsSync(configFile)) continue;
      const config = yaml.load(fs.readFileSync(configFile, "utf-8")) as Record<string, unknown>;
      if (config.guild_id === guildId) return p;
    }
    return null;
  }
}
